### PR TITLE
(feat) Added functionality to pass the Args, Commands, and ImagePullPolicy for Runner Pod

### PR DESCRIPTION
**Pull request by rahulchheda**
Signed-off-by: Rahul M Chheda <rahul.chheda@mayadata.io>

**What this PR does / why we need it**:
- This PR is for adding functionality to pass Args and Commands for the RunnerPod    
- Added fields for passing the arguments, and commands to the runner Pod for debug purposes.
- Added fields to pass imagePullPolicy

### Testcases / Considerations

- Override the go command/args 
- Use default command/args (i.e., chaosengine doesn't override) 
-  Override the ansible command/args 
- Use default command/args (i.e., chaosengine doesn't override) 
- What is the CRD validation corresponding to this? 
- BDD corresponding to this (new / changes to existing one) 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
https://github.com/litmuschaos/litmus/issues/1215

**Special notes for your reviewer**:
Example Chaos Engine Spec:
```
apiVersion: litmuschaos.io/v1alpha1
kind: ChaosEngine
metadata:
  name: engine
  namespace: litmus
spec:
  engineState: ""
  jobCleanUpPolicy: "delete"
  auxiliaryAppInfo: ""
  components:
    runner:
      image: "litmuschaos/chaos-executor:ci"
      type: "go"
      args: ["rahul"]
      command: ["m","chheda"]
      imagePullPolicy: Always
  appinfo:
    appns: litmus
    applabel: "app=nginx"
    appkind: deployment
  chaosServiceAccount: litmus
  monitoring: true
  experiments:
    - name: node-cpu-hog
```
Runner pod spec:
```
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: "2020-02-19T09:48:08Z"
  labels:
    app: engine
    chaosUID: 459a8fb6-522b-11ea-9ec5-42010a800213
  name: engine-runner
  namespace: litmus
  resourceVersion: "7420457"
  selfLink: /api/v1/namespaces/litmus/pods/engine-runner
  uid: ef30058a-52fc-11ea-9ec5-42010a800213
spec:
  containers:
  - args:
    - rahul
    command:
    - m
    - chheda
    env:
    - name: CHAOSENGINE
      value: engine
    - name: APP_LABEL
      value: app=nginx
    - name: APP_NAMESPACE
      value: litmus
    - name: EXPERIMENT_LIST
      value: node-cpu-hog
    - name: CHAOS_SVC_ACC
      value: litmus
    - name: AUXILIARY_APPINFO
    - name: CLIENT_UUID
      value: e3b34dac-6fb8-d332-f558-585906cd9fc9
    image: litmuschaos/chaos-executor:ci
    imagePullPolicy: Always
    name: chaos-runner
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: litmus-token-lrjbt
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  nodeName: gke-rahulchheda-litmus-default-pool-94f0899c-471s
  priority: 0
  restartPolicy: OnFailure
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: litmus
  serviceAccountName: litmus
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - name: litmus-token-lrjbt
    secret:
      defaultMode: 420
      secretName: litmus-token-lrjbt
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2020-02-19T09:48:08Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2020-02-19T09:48:08Z"
    message: 'containers with unready status: [chaos-runner]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2020-02-19T09:48:08Z"
    message: 'containers with unready status: [chaos-runner]'
    reason: ContainersNotReady
    status: "False"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2020-02-19T09:48:08Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: docker://7a3e519468ae1c87dc46dcd31d274d949605db4323d48a0fdad4263142f8d52c
    image: litmuschaos/chaos-executor:ci
    imageID: docker-pullable://litmuschaos/chaos-executor@sha256:aa6a834570117a6ea86a1f6c2e54ccb04137b47180c1f57aa87f7370278a6910
    lastState:
      terminated:
        containerID: docker://7a3e519468ae1c87dc46dcd31d274d949605db4323d48a0fdad4263142f8d52c
        exitCode: 127
        finishedAt: "2020-02-19T10:29:45Z"
        message: 'OCI runtime create failed: container_linux.go:345: starting container
          process caused "exec: \"m\": executable file not found in $PATH": unknown'
        reason: ContainerCannotRun
        startedAt: "2020-02-19T10:29:45Z"
    name: chaos-runner
    ready: false
    restartCount: 13
    state:
      waiting:
        message: Back-off 5m0s restarting failed container=chaos-runner pod=engine-runner_litmus(ef30058a-52fc-11ea-9ec5-42010a800213)
        reason: CrashLoopBackOff
  hostIP: 10.128.0.98
  phase: Running
  podIP: 10.80.0.2
  qosClass: BestEffort
  startTime: "2020-02-19T09:48:08Z"


```
**Checklist:**
- [ ] Fixes #<issue number>
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -7,7 +7,7 @@ IS_DOCKER_INSTALLED = $(shell which docker >> /dev/null 2>&1; echo $$?)
 PACKAGES = $(shell go list ./... | grep -v '/vendor/')
 
 # docker info
-DOCKER_REPO ?= litmuschaos
+DOCKER_REPO ?= rahulchheda1997
```
Comment (amitbhatt818) [inline Makefile]: undo this change

Comment (rahulchheda) [reply Makefile]: Yeah!

File: pkg/apis/litmuschaos/v1alpha1/chaosengine_types.go [hunk 2]
```diff
@@ -84,6 +84,10 @@ type RunnerInfo struct {
 	Image string `json:"image,omitempty"`
 	//Type of runner
 	Type string `json:"type,omitempty"`
+	//Args of runner
+	Args []string `json:"args,omitempty"`
+	//Command for runner
+	Command []string `json:"command,omitempty"`
 }
```
Comment (ksatchit) [inline pkg/apis/litmuschaos/v1alpha1/chaosengine_types.go]: imagePullPolicy can be added for runner

**Comment by ksatchit** [comment]
Have added some tests/considerations into the description. Lets factor in those as well.

File: pkg/controller/chaosengine/chaosengine_controller.go [hunk 3]
```diff
@@ -338,35 +338,65 @@ func newRunnerPodForCR(ce chaosTypes.EngineInfo) (*corev1.Pod, error) {
...
-				WithImage(engine.Instance.Spec.Components.Runner.Image).
-				WithImagePullPolicy(corev1.PullIfNotPresent)).Build()
+		WithContainerBuilder(containerForRunner).Build()
 }
 
 func newAnsibleRunnerPodForCR(engine chaosTypes.EngineInfo) (*corev1.Pod, error) {
+	containerForRunner := container.NewBuilder().
+		WithName("chaos-runner").
+		WithImage(engine.Instance.Spec.Components.Runner.Image).
+		WithImagePullPolicy(corev1.PullIfNotPresent).
```
Comment (shubhamchaudhary) [inline pkg/controller/chaosengine/chaosengine_controller.go]: Why are you adding withimagePullPolicy/Args/Command at two places?

File: pkg/apis/litmuschaos/v1alpha1/chaosengine_types.go [hunk 4]
```diff
@@ -78,12 +79,30 @@ type MonitorInfo struct {
 	Image string `json:"image"`
 }
 
+// PullPolicy describes a policy for if/when to pull a container image
+type PullPolicy string
+
+const (
+	// PullAlways means that kubelet always attempts to pull the latest image. Container will fail If the pull fails.
+	PullAlways PullPolicy = "Always"
```
Comment (shubhamchaudhary) [inline pkg/apis/litmuschaos/v1alpha1/chaosengine_types.go]: Can you please explain the use cases for this, Why are you defining them here?

Comment (rahulchheda) [reply pkg/apis/litmuschaos/v1alpha1/chaosengine_types.go]: no need, removed in new commit

File: pkg/controller/chaosengine/chaosengine_controller.go (hunk 3 above)
Comment (rahulchheda) [reply pkg/controller/chaosengine/chaosengine_controller.go]: So, first one is the default, and after that override.

File: pkg/controller/chaosengine/chaosengine_controller.go [hunk 5]
```diff
@@ -338,35 +338,62 @@ func newRunnerPodForCR(ce chaosTypes.EngineInfo) (*corev1.Pod, error) {
 }
 
 func newGoRunnerPodForCR(engine chaosTypes.EngineInfo) (*corev1.Pod, error) {
+	containerForRunner := container.NewBuilder().
+		WithEnvsNew(getChaosRunnerENV(engine.Instance, engine.AppExperiments, analytics.ClientUUID)).
+		WithName("chaos-runner").
+		WithImage(engine.Instance.Spec.Components.Runner.Image).
+		WithImagePullPolicy(corev1.PullIfNotPresent)
```
Comment (ksatchit) [inline pkg/controller/chaosengine/chaosengine_controller.go]: looks like this is the default anyways.

**Comment by ksatchit** [comment]
Can we add the validation for this please?
